### ISSUE TITLE
Sonoff Mini zeroconf/info response

### ISSUE THREAD

**Issue by olyfunt** [itead/Sonoff_Devices_DIY_Tools#34]
Here is what I receive (response body) after sending a request to zeroconf/info wanting to get the current status of the switch:

{"seq":168,"error":0,"data":"{\\"switch\\":\\"off\\",\\"startup\\":\\"stay\\",\\"pulse\\":\\"off\\",\\"pulseWidth\\":500,\\"ssid\\":\\"XXXXX\\",\\"otaUnlock\\":false}"}

Please note that the "data" attribute is actually a String value - not an object.
Why is that? Shouldn't it be rather something like below?
{"seq":168,"error":0,"data":{"switch":"off","startup":"stay","pulse":"off","pulseWidth":500,"ssid":"XXXXX","otaUnlock":false}}

I tested communication using ARC and curl.

**Comment by ZZLinvec** [comment]
Is the Sonoff Mini firmware version 3.3.0? @olyfunt

**Comment by ZZLinvec** [comment]
If so, its data field must be a json object.

**Comment by olyfunt** [comment]
The firmware version is 3.3.0. I upgraded the device right after I received it last week and before I switched to DYI mode and mounted in the wall.

**Comment by olyfunt** [comment]
I didn't notice it earlier but even DIY tool shows that data is a String value enclosed in apostrophes. 
The device shown in the image is one I unpacked and updated a few moments ago and the firmware is of course 3.3.0, the latest available.

![Sonoff_mini_info](https://user-images.githubusercontent.com/54328192/63448393-19b7c080-c43e-11e9-9f92-1537c2908335.png)

**Comment by ZZLinvec** [comment]
@olyfunt  This section displays the json data to a string, just to show the return information from getting the device data.

**Comment by olyfunt** [comment]
Why Closed? No answer has been given yet.
The question was not WHETHER "data" field is an object but WHY IT IS NOT an object.

Here is how /zeroconf/info response looks like when requested using curl:
![curl](https://user-images.githubusercontent.com/54328192/63531321-8d260480-c508-11e9-801b-a46924c76860.png)

Please copy the raw HTTP response body I provided initially:

{"seq":168,"error":0,"data":"{\\"switch\\":\\"off\\",\\"startup\\":\\"stay\\",\\"pulse\\":\\"off\\",\\"pulseWidth\\":500,\\"ssid\\":\\"XXXXX\\",\\"otaUnlock\\":false}"}

paste it in any online JSON parser and check for yourself.
![json_parse](https://user-images.githubusercontent.com/54328192/63530331-8c8c6e80-c506-11e9-94be-9cf1c5f57645.png)

When a piece of software receives such a response and parse it, it gets an object with 3 fields like below:
{
   "seq": number,
   "error": number,
   "data": string
}
Despite the things you wrote here:
https://github.com/itead/Sonoff_Devices_DIY_Tools/blob/master/SONOFF%20DIY%20MODE%20Protocol%20Doc%20v1.4.md#8-get-device-info
forget about reading e.g. data.switch to check whether it is "on" or "off".
This is because "data" field is just a plain text only containing JSON definition. It needs to be parsed yet in order to get an object and read the fields inside.

I can deal with it if I have to. I was only curious if this a bit strange solution serves any purpose.
Do you see my point?

**Comment by ZZLinvec** [comment]
sorry，This is my fault. I have rechecked the information returned by the device. At present, this information is indeed of type string.

**Comment by ZZLinvec** [comment]
This is a mistake in the firmware, we will improve it.

**Comment by olyfunt** [comment]
Thank you @ZZLinvec

**Comment by JAMESBOWLER** [comment]
I found the same issue I have it running in openhab. I just transform the data object into anthrer string being a json one and then use that as json data to put all the data into relevant item.

`
 if (UpdateResult.contains('"error":0')) { 
        JSONdata = transform("JSONPATH", "$.data", UpdateResult)      
       // logInfo("Sonoff Poll JSONdata", JSONdata)


      if (JSONdata != Last_update) {
      
        update_switch_state = transform("JSONPATH", "$.switch", JSONdata)
`

**Comment by alexbk66** [comment]
Yeah, works for me:
`SonoffData data = JsonConvert.DeserializeObject<SonoffData>(data);`

**Comment by alexbk66** [comment]
@ZZLinvec - if you now change the firmware - you will break existing code.  Is it worths it?
And I wonder - do you follow proper software development procedures, do you do testing?

**Comment by frenck** [comment]
From the consumer endpoint, a change for this can be mitigated pretty easily by checking the type of the `data` payload.

IMHO, this should be fixed ASAP. The longer you guys wait on fixing this, the harder it becomes.

**Comment by alexbk66** [comment]
> From the consumer endpoint, a change for this can be mitigated pretty easily by checking the type of the `data` payload.

Yeah, "pretty easy"?  But existing software will have to be updated **before** the firmware update.

**Comment by Daniel-zhan-itead** [comment]
We will update zeroconf/info response body as object but not strings, in the next firmware.

**Comment by frenck** [comment]
Thanks, @Daniel-zhan-itead

**Comment by alexbk66** [comment]
> We will update zeroconf/info response body as object but not strings, in the next firmware.

Hi Daniel @Daniel-zhan-itead, just wondering - are you planning to implement OTA FW update for DIY?  Or we have to go off DIY mode to be able to use the eWeLink app to update FW and then go through the complicate pairing procedure to go back to DIY mode?

**Comment by rogerjames99** [comment]
Is there any update on a timescale for releasing this fix? The current version appears to break the openhab JSONPATH transform. Yes, what is returned by the device is a valid piece of JSON but the "data" field is an object of type string, not a nested piece of JSON. Handling this puts unecessary extra processing load on upstream devices which may themselves be embedded devices with limited resources. Fixing this should be done as a matter of urgency. It will reduce the complexity of the code on the device, surely it is more complex to do this double encoding rather than encoding the whole thing as a single valid piece of JSON.
